File: atsa/src/main/java/jp/kaiz/atsassistmod/controller/TrainController.java
package jp.kaiz.atsassistmod.controller;

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity;
import jp.kaiz.atsassistmod.controller.trainprotection.TrainProtection;
import jp.kaiz.atsassistmod.controller.trainprotection.TrainProtectionType;
import jp.kaiz.atsassistmod.rtm.RtmTrains;
import net.minecraft.world.phys.Vec3;

import java.util.ArrayList;
import java.util.List;

public class TrainController {
    private final List<SpeedOrder> speedOrderList = new ArrayList<>();
    private final List<Integer> speedLimit = new ArrayList<>();
    private int maxSpeed = 0;

    private boolean ATO = false;
    private boolean acceleratorControlling = false;
    public final TASCController tascController = new TASCController();
    private boolean brakingControlling = false;

    private TrainProtection tp;
    private TrainProtectionType tpType = TrainProtectionType.NONE;

    private TrainEntity train;

    private Vec3 coordinates;

    private final int savedEntityID;

    private byte controllerNotchA = -1, controllerNotchB = 1;
    private boolean controllerControl = false;
    private boolean emergencyBrake = false;
    private boolean manualDrive = false;

    public static final TrainController NULL = new TrainController();

    public TrainController() {
        this.savedEntityID = -1;
        this.setTrainProtection(TrainProtectionType.NONE);
    }

    public TrainController(TrainEntity train) {
        this.savedEntityID = train.getId();
        this.train = train;
        this.setTrainProtection(TrainProtectionType.NONE);
    }

    /** Keeps the controller pointed at the live train instance for its formation. */
    public void bind(TrainEntity train) {
        this.train = train;
    }

    public TrainEntity getTrain() {
        return this.train;
    }

    public void setEB() {
        this.emergencyBrake = true;
        if (this.train != null) {
            this.train.setNotch(-8);
        }
    }

    public void setManualDrive(boolean manualDrive) {
        this.manualDrive = manualDrive;
    }

    public boolean isManualDrive() {
        return this.manualDrive;
    }

    public void setControllerNotch(byte notch) {
        this.controllerControl = true;
        if (notch > 0) {
            this.controllerNotchA = notch;
            this.controllerNotchB = 1;
        } else if (notch < 0) {
            this.controllerNotchA = 0;
            this.controllerNotchB = notch;
        } else {
            this.controllerNotchA = 0;
            this.controllerNotchB = 1;
        }
    }

    public int getSavedEntityID() {
        return savedEntityID;
    }

    public void addSpeedOrder(SpeedOrder speedOrder) {
        this.speedOrderList.add(speedOrder);
    }

    public void setMaxSpeed(int maxSpeed) {
        this.maxSpeed = maxSpeed;
    }

    public void removeSpeedLimit() {
        if (!this.speedLimit.isEmpty()) {
            this.speedLimit.remove(0);
        }
    }

    public void removeAllSpeedLimit() {
        if (!this.speedLimit.isEmpty()) {
            this.speedLimit.clear();
        }
    }

    public int getSpeedLimit() {
        return this.speedLimit.stream().mapToInt(v -> v).min().orElse(Integer.MAX_VALUE);
    }

    public int getATOSpeedLimit() {
        int minLimit = this.speedLimit.stream().mapToInt(v -> v).min().orElse(this.maxSpeed);
        return Math.min(this.getTrainProtectionSpeedLimit(), Math.min(minLimit, this.maxSpeed));
    }

    public int getTrainProtectionSpeedLimit() {
        return this.tp.getDisplaySpeed();
    }

    public void enableATO(int speed) {
        this.ATO = true;
        this.maxSpeed = speed;
    }

    public void disableATO() {
        this.ATO = false;
    }

    public boolean isATO() {
        return ATO;
    }

    public void setTrainProtection(TrainProtectionType type) {
        this.tpType = type;
        this.tp = type.newInstance();
    }

    public TrainProtectionType getTrainProtectionType() {
        return this.tp.getType();
    }

    /** Tick processing (server side). */
    public void onUpdate() throws Exception {
        if (train == null) {
            return;
        }

        double movedDistance = this.getMovedDistance();
        float speedH = RtmTrains.speedKmh(train);

        List<Integer> brakeNotch = new ArrayList<>();
        List<Integer> acceleratorNotch = new ArrayList<>();

        List<SpeedOrder> removeList = new ArrayList<>();
        this.speedOrderList.forEach(speedOrder -> {
            if (speedOrder.isEnable()) {
                this.speedLimit.add(speedOrder.getTargetSpeed());
                removeList.add(speedOrder);
            } else {
                speedOrder.moveDistance(movedDistance);
                if (speedOrder.isAutoBrake() || (this.ATO && !this.isManualDrive())) {
                    brakeNotch.add(speedOrder.getNeedNotch(speedH));
                }
            }
        });
        this.speedOrderList.removeAll(removeList);

        if (speedH > this.getSpeedLimit()) {
            float overSpeed = speedH - this.getSpeedLimit();
            if (overSpeed < 5f) {
                brakeNotch.add(-4);
            } else {
                brakeNotch.add(-7);
            }
        } else {
            if (this.ATO && !this.isManualDrive()) {
                if (!this.acceleratorControlling) {
                    if ((this.getATOSpeedLimit() - speedH) > 10) {
                        acceleratorNotch.add(5);
                    } else if (speedH == 0) {
                        acceleratorNotch.add(5);
                    }
                } else if (this.getATOSpeedLimit() - speedH < 2) {
                    acceleratorNotch.add(0);
                }
            }
        }

        this.tascController.changeTargetDistance(movedDistance);
        if (this.tascController.isEnable()) {
            int needNotch = this.tascController.getNeedNotch(speedH);
            if (this.tascController.isBreaking()) {
                this.disableATO();
                if (!this.isManualDrive()) {
                    brakeNotch.add(needNotch);
                }
            }
        }

        this.tp.onTick(train, movedDistance);
        int notch = this.tp.getNotch(speedH);
        brakeNotch.add(notch);

        if (this.emergencyBrake) {
            int notchLevel = this.train.getNotch();
            if (notchLevel != -8) {
                this.emergencyBrake = false;
                this.brakingControlling = false;
            } else {
                return;
            }
        }

        int minBrakeNotch = brakeNotch.stream().mapToInt(v -> v).min().orElse(1);

        if (RtmTrains.rider(this.train) == null) {
            this.controllerNotchA = -1;
            this.controllerNotchB = 1;
            if (this.controllerControl) {
                this.controllerControl = false;
                this.brakingControlling = false;
            }
        } else {
            if (this.controllerControl) {
                minBrakeNotch = Math.min(minBrakeNotch, this.controllerNotchB);
            }
        }

        if (minBrakeNotch > 0) {
            int maxAcceleratorNotch = acceleratorNotch.stream().mapToInt(v -> v).max().orElse(-1);

            if (RtmTrains.rider(this.train) != null) {
                if (this.controllerControl) {
                    maxAcceleratorNotch = Math.max(maxAcceleratorNotch, this.controllerNotchA);
                }
            }

            if (maxAcceleratorNotch < 0) {
                if (this.brakingControlling) {
                    this.brakingControlling = false;
                    this.train.setNotch(0);
                }
            } else if (maxAcceleratorNotch == 0) {
                this.brakingControlling = false;
                if (!this.controllerControl || this.controllerNotchA != 0) {
                    this.acceleratorControlling = false;
                }
                this.train.setNotch(0);
            } else {
                this.brakingControlling = false;
                this.acceleratorControlling = true;
                this.train.setNotch(maxAcceleratorNotch);
            }
        } else if (minBrakeNotch == 0) {
            if (this.tascController.isEnable() && this.tascController.isStopPosition()) {
                this.brakingControlling = false;
                this.ATO = false;
                if (speedH <= 0F) {
                    this.tascController.disable();
                    if (!this.isManualDrive()) {
                        return;
                    }
                }
            }
            this.acceleratorControlling = false;
            this.brakingControlling = false;
            this.train.setNotch(0);
        } else {
            this.acceleratorControlling = false;
            if (this.tascController.isEnable() && this.tascController.isStopPosition()) {
                this.brakingControlling = false;
                this.ATO = false;
                if (speedH <= 0F) {
                    this.tascController.disable();
                    if (!this.isManualDrive()) {
                        return;
                    }
                }
            }
            this.brakingControlling = true;
            this.train.setNotch(minBrakeNotch);
        }

        if (this.tascController.isEnable() && this.tascController.isStopPosition()) {
            this.ATO = false;
            if (this.isManualDrive()) {
                this.tascController.disable();
            }
        }
    }

    private double getMovedDistance() {
        Vec3 now = RtmTrains.pos(train);
        if (this.coordinates == null) {
            this.coordinates = now;
            return 0d;
        }
        Vec3 old = this.coordinates;
        this.coordinates = now;
        // horizontal moved distance (no gradient), matching the original
        double dx = old.x - now.x;
        double dz = old.z - now.z;
        return Math.sqrt(dx * dx + dz * dz);
    }

    // ---- accessors used by HUD/network sync ----
    public boolean isTASCEnable() { return this.tascController.isEnable(); }
    public boolean isEmergencyBrake() { return this.emergencyBrake; }
    public TrainProtectionType getTpType() { return this.tpType; }
}

File: atsa/src/main/kotlin/jp/kaiz/atsassistmod/controller/TrainController.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package jp.kaiz.atsassistmod.controller

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity
import jp.kaiz.atsassistmod.controller.trainprotection.TrainProtection
import jp.kaiz.atsassistmod.controller.trainprotection.TrainProtectionType
import jp.kaiz.atsassistmod.rtm.RtmTrains
import net.minecraft.world.phys.Vec3
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

open class TrainController {
    private val speedOrderList = ArrayList<SpeedOrder>()
    private val speedLimit = ArrayList<Int>()
    private var maxSpeed = 0

    private var ato = false
    private var acceleratorControlling = false

    @JvmField
    val tascController = TASCController()

    private var brakingControlling = false

    private lateinit var tp: TrainProtection
    private var tpType = TrainProtectionType.NONE
    private var train: TrainEntity? = null
    private var coordinates: Vec3? = null
    private val savedEntityID: Int
    private var controllerNotchA: Byte = -1
    private var controllerNotchB: Byte = 1
    private var controllerControl = false
    private var emergencyBrake = false
    private var manualDrive = false

    constructor() {
        savedEntityID = -1
        setTrainProtection(TrainProtectionType.NONE)
    }

    constructor(train: TrainEntity) {
        savedEntityID = train.id
        this.train = train
        setTrainProtection(TrainProtectionType.NONE)
    }

    /** Keeps the controller pointed at the live train instance for its formation. */
    fun bind(train: TrainEntity) {
        this.train = train
    }

    fun getTrain(): TrainEntity? = train

    fun setEB() {
        emergencyBrake = true
        train?.notch = -8
    }

    fun setManualDrive(manualDrive: Boolean) {
        this.manualDrive = manualDrive
    }

    fun isManualDrive(): Boolean = manualDrive

    fun setControllerNotch(notch: Byte) {
        controllerControl = true
        when {
            notch > 0 -> {
                controllerNotchA = notch
                controllerNotchB = 1
            }
            notch < 0 -> {
                controllerNotchA = 0
                controllerNotchB = notch
            }
            else -> {
                controllerNotchA = 0
                controllerNotchB = 1
            }
        }
    }

    fun getSavedEntityID(): Int = savedEntityID

    fun addSpeedOrder(speedOrder: SpeedOrder) {
        speedOrderList.add(speedOrder)
    }

    fun setMaxSpeed(maxSpeed: Int) {
        this.maxSpeed = maxSpeed
    }

    fun removeSpeedLimit() {
        if (speedLimit.isNotEmpty()) {
            speedLimit.removeAt(0)
        }
    }

    fun removeAllSpeedLimit() {
        if (speedLimit.isNotEmpty()) {
            speedLimit.clear()
        }
    }

    fun getSpeedLimit(): Int = speedLimit.minOrNull() ?: Int.MAX_VALUE

    fun getATOSpeedLimit(): Int {
        val minLimit = speedLimit.minOrNull() ?: maxSpeed
        return min(getTrainProtectionSpeedLimit(), min(minLimit, maxSpeed))
    }

    fun getTrainProtectionSpeedLimit(): Int = tp.getDisplaySpeed()

    fun enableATO(speed: Int) {
        ato = true
        maxSpeed = speed
    }

    fun disableATO() {
        ato = false
    }

    fun isATO(): Boolean = ato

    fun setTrainProtection(type: TrainProtectionType) {
        tpType = type
        tp = type.newInstance()
    }

    fun getTrainProtectionType(): TrainProtectionType = tp.getType()

    /** Tick processing (server side). */
    @Throws(Exception::class)
    fun onUpdate() {
        val train = train ?: return
        val movedDistance = getMovedDistance()
        val speedH = RtmTrains.speedKmh(train)

        val brakeNotch = ArrayList<Int>()
        val acceleratorNotch = ArrayList<Int>()
        val removeList = ArrayList<SpeedOrder>()

        speedOrderList.forEach { speedOrder ->
            if (speedOrder.isEnable()) {
                speedLimit.add(speedOrder.getTargetSpeed())
                removeList.add(speedOrder)
            } else {
                speedOrder.moveDistance(movedDistance)
                if (speedOrder.isAutoBrake() || (ato && !isManualDrive())) {
                    brakeNotch.add(speedOrder.getNeedNotch(speedH))
                }
            }
        }
        speedOrderList.removeAll(removeList)

        if (speedH > getSpeedLimit()) {
            val overSpeed = speedH - getSpeedLimit()
            if (overSpeed < 5f) {
                brakeNotch.add(-4)
            } else {
                brakeNotch.add(-7)
            }
        } else if (ato && !isManualDrive()) {
            if (!acceleratorControlling) {
                if (getATOSpeedLimit() - speedH > 10 || speedH == 0f) {
                    acceleratorNotch.add(5)
                }
            } else if (getATOSpeedLimit() - speedH < 2) {
                acceleratorNotch.add(0)
            }
        }

        tascController.changeTargetDistance(movedDistance)
        if (tascController.isEnable()) {
            val needNotch = tascController.getNeedNotch(speedH)
            if (tascController.isBreaking()) {
                disableATO()
                if (!isManualDrive()) {
                    brakeNotch.add(needNotch)
                }
            }
        }

        tp.onTick(train, movedDistance)
        brakeNotch.add(tp.getNotch(speedH))

        if (emergencyBrake) {
            val notchLevel = train.notch
            if (notchLevel != -8) {
                emergencyBrake = false
                brakingControlling = false
            } else {
                return
            }
        }

        var minBrakeNotch = brakeNotch.minOrNull() ?: 1

        if (RtmTrains.rider(train) == null) {
            controllerNotchA = -1
            controllerNotchB = 1
            if (controllerControl) {
                controllerControl = false
                brakingControlling = false
            }
        } else if (controllerControl) {
            minBrakeNotch = min(minBrakeNotch, controllerNotchB.toInt())
        }

        if (minBrakeNotch > 0) {
            var maxAcceleratorNotch = acceleratorNotch.maxOrNull() ?: -1

            if (RtmTrains.rider(train) != null && controllerControl) {
                maxAcceleratorNotch = max(maxAcceleratorNotch, controllerNotchA.toInt())
            }

            when {
                maxAcceleratorNotch < 0 -> {
                    if (brakingControlling) {
                        brakingControlling = false
                        train.notch = 0
                    }
                }
                maxAcceleratorNotch == 0 -> {
                    brakingControlling = false
                    if (!controllerControl || controllerNotchA.toInt() != 0) {
                        acceleratorControlling = false
                    }
                    train.notch = 0
                }
                else -> {
                    brakingControlling = false
                    acceleratorControlling = true
                    train.notch = maxAcceleratorNotch
                }
            }
        } else if (minBrakeNotch == 0) {
            if (tascController.isEnable() && tascController.isStopPosition()) {
                brakingControlling = false
                ato = false
                if (speedH <= 0f) {
                    tascController.disable()
                    if (!isManualDrive()) {
                        return
                    }
                }
            }
            acceleratorControlling = false
            brakingControlling = false
            train.notch = 0
        } else {
            acceleratorControlling = false
            if (tascController.isEnable() && tascController.isStopPosition()) {
                brakingControlling = false
                ato = false
                if (speedH <= 0f) {
                    tascController.disable()
                    if (!isManualDrive()) {
                        return
                    }
                }
            }
            brakingControlling = true
            train.notch = minBrakeNotch
        }

        if (tascController.isEnable() && tascController.isStopPosition()) {
            ato = false
            if (isManualDrive()) {
                tascController.disable()
            }
        }
    }

    private fun getMovedDistance(): Double {
        val train = train ?: return 0.0
        val now = RtmTrains.pos(train)
        val old = coordinates
        if (old == null) {
            coordinates = now
            return 0.0
        }
        coordinates = now
        val dx = old.x - now.x
        val dz = old.z - now.z
        return sqrt(dx * dx + dz * dz)
    }

    fun isTASCEnable(): Boolean = tascController.isEnable()

    fun isEmergencyBrake(): Boolean = emergencyBrake

    fun getTpType(): TrainProtectionType = tpType

    companion object {
        @JvmField
        val NULL = TrainController()
    }
}
